apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClientImpl.SingleChannel.cs

```diff
@@ -1109,61 +1109,121 @@ private void UpdateReceiptModAckStatus(IEnumerable<string> ids, bool? status)
+                _logger = logger;
+                _backoffTiming = backoffTiming;
                 _disconnectBackoff = disconnectBackoff;
+                _exceptions = new();
+                _firstExceptionTimestamp = null;
                 Backoff = null;
             }
 
             /// <summary>
-            /// Checks whether the given exception should be retried, without updating any state.
-            /// </summary>
-            public bool ShouldRetry(Exception exception) =>
-                exception?.As<RpcException>()?.IsRecoverable() ?? false;
-
-            /// <summary>
-            /// Records the given exception in the retry state, updating the backoff accordingly.
+            /// Checks whether the given exception should be retried, updating the state including the next backoff.
             /// </summary>
-            public void OnFailure(Exception exception)
+            internal bool RecordFailureAndCheckForRetry(Exception exception)
             {
-                _firstExceptionTimestamp ??= _clock.GetCurrentDateTimeUtc();
-                if (exception?.As<RpcException>() is RpcException rpcException && _exceptions.Count < 100)
+                var now = _clock.GetCurrentDateTimeUtc();
+                _firstExceptionTimestamp ??= now;
+                // This won't be used if we decide not to retry anyway.
+                Backoff = _backoffTiming.NextBackoff(Backoff ?? TimeSpan.Zero);
+
+                // Don't retry any non-RpcExceptions
+                if (exception.As<RpcException>() is not RpcException rpcEx)
+                {
+                    return false;
+                }
+                if (_exceptions.Count < ExceptionLimit)
                 {
-                    _exceptions.Add(rpcException);
+                    _exceptions.Add(rpcEx);
                 }
-                Backoff = _retrySettings.NextBackoff(Backoff ?? TimeSpan.Zero);
+
+                var code = rpcEx.StatusCode;
+
+                // If the server has just dropped the stream, but it's after a sufficiently long time, we can
+                // deem that to be successful and retry with no backoff.
+                if (code == StatusCode.Unavailable && (now - _currentStartTimestamp) >= s_streamingPullSuccessThreshold)
+                {
+                    _logger?.LogDebug("Pull stream terminated with no messages, but after success assumption threshold time. Retrying with no backoff.");
+                    OnSuccess();
+                    return true;
+                }
+
+                // If the exception isn't generally recoverable, don't retry.
+                if (!rpcEx.IsRecoverable())
+                {
+                    return false;
+                }
+
+                // If the exception was a failure due to auth, and we've seen some before, don't retry.
+                if (IsAuthException(rpcEx))
+                {
+                    var count = _exceptions.Count(IsAuthException);
```

Comment:
Corner case likely as we've been failing anyway, but shouldn't the auth exceptions be consecutive for us to stop retrying? If there was an Auth exception and then 10 other exceptions, and then an Auth exception, etc. it might be that the auth error is really transient (and there's something else going on).

Reply:
Discussed offline: we're going to leave this, as if there's a mix of errors it means things are pretty bad and we're unlikely to recover properly.